Update SPI HAL API and example implementations

mprse opened the pull request:
### Description

This PR provides the following changes:
 - NUCLEO_F429ZI, K66F, K64F SPI driver fixes,
 - spi_api.h: Update defined/undefined behavior,
 - NUCLEO_F429ZI, K66F Spi driver: update capabilities,
 - Rework spi_get_capabilities() function.

Please check commit description for more info.

### Pull request type

    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@ithinuel 
@c1728p9

mprse committed:
NUCLEO_F429ZI, K66F, K64F SPI driver fixes.

While testing usnig FPGA test shield the following problems have been found:
NUCLEO_F429ZI, K66F, K64F:
- Function spi_get_capabilities() should consider the `ssel` pin when evaluation the `support_slave_mode` capability. If the given `ssel` pin cannot be managed by hardware in slave mode, `support_slave_mode` should be false.

K66F, K64F:
- `set_frequency()` cannot be called before `spi_format()`.
- `spi_transfer()` provides currently special handling for one symbol transfer, which causes problems when one symbols is being transmitted.
- `spi_init()` Chip Select pin must be specified (NOT NC) in slave mode (missing assert).
- `spi_init()`/`spi_free()` pins function is not reset by `spi_free()`.

NUCLEO_F429ZI:
- `spi_transfer()` in half duplex mode setting transmission timeout(`HAL_SPI_Transmit()`) causes transmission issues which are not found when unlimited timeout is set.

This commit fixes above problems.
Add checks if pin is defined before deinitialization.

mprse committed:
spi_api.h: Update defined/undefined behavior

mprse committed:
NUCLEO_F429ZI, K66F Spi driver: update capabilities

NUCLEO_F429ZI:
- Chip Select cannot be managed by hardware (there is no possibility to specify delay between Chip Select assertion and first clock edge / last clock edge and Chip select de-assertion).

NUCLEO_K66F:
- Chip Select can be managed by hardware.
- Half-Duplex mode is not supported.

mprse committed:
Rework spi_get_capabilities() function

Perform the following changes:
 - add param which specifies if we want to retrieve capabilities of SPI slave or SPI master,
 - add information about supported: clock modes, bit orders, async mode, delay between symbols(slave mode),
 - update docs.

c1728p9 commented on docs/design-documents/hal/0000-spi-overhaul.md:
```diff
@@ -98,14 +98,15 @@ typedef struct {
-    uint32_t    symbol_length;
+    /** Each bit represents the corresponding word length. lsb => 1bit, msb => 32bit. */
+    uint32_t    word_length;
+    uint16_t    slave_delay_between_symbols_ns; /**< specifies required number of ns between transmission of successive symbols in slave mode. */
```
Would this be more appropriate as `maximum_effective_frequency`? This is the fastest a SPI master using DMA will be able to send data to a slave without losing bytes.

mprse replied:
`maximum_effective_frequency` would be the same as `maximum_frequency` for spi slave. 

The problem is that our implementation for synchronous transfer doesn't use DMA. On `K66F` we got TX, RX fifos and there is no problem on the slave side to handle max frequency reported by device capabilities (no delay between symbols is required). 
On `NUCLEO_F429ZI` symbols are transmitted one by one by writing/reading successive symbols to/from the TX/RX registers. Because of that max effective frequency for the slave is ~1 MHz. I wanted to set `maximum_frequency` for the slave to 1 MHz, but this is not consistent with the following requirement:
`The supported frequency range must include the range [0.2..2] MHz.`

mprse committed:
Update basic SPI test.

Perform the following changes:
 - Adapt usage of spi_get_capabilities() (API has changed),
 - Make use of pinmap extension instead using hardcoded pins,
 - SPISLAVE -> DEVICE_SPISLAVE

mprse committed:
pinmap_find_peripheral_pins(): use whitelist and blacklist only if defined

Add check if whitelist/blacklist is defined before checking if specified pin is on the list.
This will extend usability of the function (using whitelist/blacklist will be optional).

For example in case of basic SPI test we need any valid pins configuration of any peripheral.

mprse committed:
K64F, K66F, NUCLEO_F429ZI: Remove redundent spi pins definitions.

Instead of hardcoded spi pins pinmpa extension is used in tests.

mprse committed:
tests-mbed_hal-spi_com: update usage of spi_get_capabilities() and specify pins for testing in the test header.

mprse committed:
SPI IceTea Com test: update usage of spi_get_capabilities()

mprse committed:
Disable SERIAL_FC for `ARM_IOTSS_BEID, ARM_MPS2 family`

Looks like listed boards have SERIAL_FC feature enabled, but does not provide pins definition for `PinMap_UART_CTS`, `PinMap_UART_RTS`.

This should fix the following failures:
ARM_IOTSS_BEID
ARM_MPS2_M0
ARM_MPS2_M3
ARM_MPS2_M4
ARM_MPS2_M7
[Error] serial_api.c@402,12: use of undeclared identifier 'PinMap_UART_RTS'

mprse committed:
UHURU_RAVEN: Disable `SPIF` component.

SPI is disabled for this target on the spi feature branch.
